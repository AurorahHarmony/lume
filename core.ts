import type {
  DeepPartial,
  DenoConfig,
  DenoConfigResult,
  ImportMap,
} from "./core/utils.ts";
import type {
  ComponentsOptions,
  default as Site,
  Plugin,
  ServerOptions,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  WatcherOptions,
} from "./core/site.ts";

import type {
  default as Events,
  Event,
  EventListener,
  EventOptions,
} from "./core/events.ts";

import type { Archetype } from "./cli/create.ts";
import type { default as Scripts, ScriptOrFunction } from "./core/scripts.ts";

import type { default as FS, Entry, Loader } from "./core/fs.ts";
import type Searcher from "./core/searcher.ts";
import type Writer from "./core/writer.ts";
import type DataLoader from "./core/data_loader.ts";
import type {
  Component,
  Components,
  default as ComponentLoader,
} from "./core/component_loader.ts";
import type {
  Content,
  Data,
  MergeStrategy,
  Page,
  Src,
  StaticFile,
} from "./core/file.ts";
import type {
  ComponentFunction,
  default as Source,
  ProxyComponents,
} from "./core/source.ts";
import type {
  default as Renderer,
  Engine,
  Helper,
  HelperOptions,
} from "./core/renderer.ts";
import type {
  default as Processors,
  Extensions,
  MultiProcessor,
  Processor,
} from "./core/processors.ts";
import type { default as Scopes, ScopeFilter } from "./core/scopes.ts";
import type { default as Formats, Format } from "./core/formats.ts";
import type {
  default as Server,
  Middleware,
  RequestHandler,
  ServerEvent,
  ServerEventType,
} from "./core/server.ts";
import type {
  default as Watcher,
  WatchEvent,
  WatchEventType,
} from "./core/watcher.ts";

// Plugins types
import type { PaginationInfo, Paginator } from "./plugins/paginate.ts";
import type { Transformation } from "./plugins/imagick.ts";
import type { MetaData } from "./plugins/metas.ts";
import type { Children } from "./plugins/jsx.ts";
import type { SourceMap } from "./plugins/source_maps.ts";

/** The method that installs a plugin */
type PluginSetup = (options: unknown) => Plugin;

export type {
  Archetype,
  Component,
  ComponentFunction,
  ComponentLoader,
  Components,
  ComponentsOptions,
  Content,
  Data,
  DataLoader,
  DeepPartial,
  DenoConfig,
  DenoConfigResult,
  Engine,
  Entry,
  Event,
  EventListener,
  EventOptions,
  Events,
  Extensions,
  Format,
  Formats,
  FS,
  Helper,
  HelperOptions,
  ImportMap,
  Loader,
  MergeStrategy,
  Middleware,
  MultiProcessor,
  Page,
  Plugin,
  PluginSetup,
  Processor,
  Processors,
  ProxyComponents,
  Renderer,
  RequestHandler,
  ScopeFilter,
  Scopes,
  ScriptOrFunction,
  Scripts,
  Searcher,
  Server,
  ServerEvent,
  ServerEventType,
  ServerOptions,
  Site,
  SiteEvent,
  SiteEventType,
  SiteOptions,
  Source,
  SourceMap,
  Src,
  StaticFile,
  Watcher,
  WatcherOptions,
  WatchEvent,
  WatchEventType,
  Writer,
};

export interface PageData extends Data {
  /** The language(s) of the page */
  lang?: string | string[];

  /** The title of the page */
  title?: string;

  /** The page url */
  url: string | false;

  /** The page reference */
  page: Page;

  /**
   * The available components
   * @see https://lume.land/docs/core/components/
   */
  // deno-lint-ignore no-explicit-any
  comp?: any;

  /**
   * The paginator helper
   * @see https://lume.land/plugins/paginate/
   */
  paginate: Paginator;

  /**
   * The pagination info
   * @see https://lume.land/plugins/paginate/
   */
  pagination?: PaginationInfo;

  /**
   * The pagination result
   * @see https://lume.land/plugins/paginate/
   */
  results?: Page[];

  /**
   * Image transformations
   * @see https://lume.land/plugins/imagick/
   */
  imagick?: Transformation | Transformation[];

  /**
   * Meta elements
   * @see https://lume.land/plugins/metas/
   */
  metas?: MetaData;

  /**
   * Decap CMS configuration
   * @see https://lume.land/plugins/decap_cms/
   */
  decap_cms?: Record<string, unknown>;

  /**
   * The searcher helper
   * @see https://lume.land/plugins/search/
   */
  search: Searcher;

  /**
   * The source map data (if it's an asset)
   */
  sourceMap?: SourceMap;

  /**
   * Alternate pages (for languages)
   */
  alternates?: PageData[];

  /**
   * The JSX children elements
   * @see https://lume.land/plugins/jsx/
   */
  children?: Children;
}

export interface PageHelpers {
  /** @see https://lume.land/plugins/attributes/ */
  attr: Helper;

  /** @see https://lume.land/plugins/attributes/ */
  class: Helper;

  /** @see https://lume.land/plugins/date/ */
  date: Helper;

  /** @see https://lume.land/plugins/liquid/ */
  liquid: Helper;

  /** @see https://lume.land/plugins/markdown/ */
  md: Helper;

  /** @see https://lume.land/plugins/nunjucks/ */
  njk: Helper;

  /** @see https://lume.land/plugins/postcss/ */
  postcss: Helper;

  /** @see https://lume.land/plugins/pug/ */
  pug: Helper;

  /** @see https://lume.land/plugins/slugify_urls/ */
  slugify: Helper;

  /** @see https://lume.land/plugins/terser/#the-terser-filter */
  terser: Helper;

  /** @see https://lume.land/plugins/url/#url-filter */
  url: Helper;

  /** @see https://lume.land/plugins/url/#htmlurl-filter */
  htmlUrl: Helper;

  [key: string]: Helper | undefined;
}
