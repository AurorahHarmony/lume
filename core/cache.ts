// deno-lint-ignore-file no-explicit-any

import { emptyDir, ensureDir } from "../deps/fs.ts";
import { posix } from "../deps/path.ts";
import { crypto } from "../deps/crypto.ts";
import { encodeHex } from "../deps/hex.ts";

export interface Options {
  /** The folder to load the files from */
  folder?: string;
}

/**
 * Class to cache the content transformations (like imagick manipulations)
 */
export default class Cache {
  #folder: string;

  constructor(options: Options = {}) {
    this.#folder = options.folder || "_cache";
  }

  async set(
    content: string | Uint8Array,
    key: any,
    result: string | Uint8Array,
  ): Promise<void> {
    const [dir, file] = await paths(content, key);

    await ensureDir(posix.join(this.#folder, dir));

    if (typeof result === "string") {
      await Deno.writeTextFile(posix.join(this.#folder, dir, file), result);
    } else {
      await Deno.writeFile(posix.join(this.#folder, dir, file), result);
    }
  }

  async get(content: any, key: any): Promise<Uint8Array | undefined> {
    const [dir, file] = await paths(content, key);

    try {
      return await Deno.readFile(posix.join(this.#folder, dir, file));
    } catch {
      // Ignore
    }
  }

  async getText(content: any, key: any): Promise<string | undefined> {
    const [dir, file] = await paths(content, key);

    try {
      return await Deno.readTextFile(posix.join(this.#folder, dir, file));
    } catch {
      // Ignore
    }
  }

  async clear(): Promise<void> {
    await emptyDir(this.#folder);
  }
}

function paths(
  content: string | Uint8Array,
  key: any,
): Promise<[string, string]> {
  return Promise.all([
    hash(content),
    hash(JSON.stringify(key)),
  ]);
}

async function hash(content: string | Uint8Array): Promise<string> {
  const hash = await crypto.subtle.digest(
    "MD5",
    typeof content === "string" ? new TextEncoder().encode(content) : content,
  );

  return encodeHex(hash);
}
