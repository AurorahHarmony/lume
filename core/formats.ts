import { Exception } from "./errors.ts";

import type { Engine, Loader } from "../core.ts";

export interface Format {
  /** The file extension for this format */
  ext: string;

  /** The file loader used for this format */
  pageLoader?: Loader;

  /** Loader for _data files in this format */
  dataLoader?: Loader;

  /** Loader for _components files in this format */
  componentLoader?: Loader;

  /**
   * The template engines used to render this format
   * Used to render the page and components
   */
  engines?: Engine[];

  /**
   * This is used to distinguish between pages that output html files (like index.njk -> index.html)
   * and pages that output assets (like styles.css -> styles.css).
   */
  asset?: boolean;

  /** A custom path for includes */
  includesPath?: string;

  /** Whether this file must be copied instead loaded */
  copy?: boolean | ((path: string) => string);
}

/** Class to store loaders, engines and other stuff related with different formats */
export default class Formats {
  entries = new Map<string, Format>();

  get size(): number {
    return this.entries.size;
  }

  /** Assign a value to a extension */
  set(format: Format, override = true): void {
    const { ext } = format;
    const existing = this.entries.get(ext);

    if (existing) {
      if (override) {
        this.entries.set(ext, { ...existing, ...format });
      } else {
        this.entries.set(ext, { ...format, ...existing });
      }
      return;
    }

    // Simple extension (.ts, .js, .json)
    if (ext.match(/^\.\w+$/)) {
      this.entries.set(ext, format);
      return;
    }

    // Chained extension (.tmpl.js) goes first
    if (ext.match(/^\.\w+\.\w+$/)) {
      const entries = Array.from(this.entries.entries());
      entries.unshift([ext, format]);
      this.entries = new Map(entries);
      return;
    }

    throw new Exception(
      "Invalid file extension. It must start with '.'",
      { ext },
    );
  }

  /** Returns a format by extension */
  get(extension: string): Format | undefined {
    return this.entries.get(extension);
  }

  /** Delete a format */
  delete(extension: string): void {
    this.entries.delete(extension);
  }

  /** Returns if a format exists */
  has(extension: string): boolean {
    return this.entries.has(extension);
  }

  /** Search and return the associated format for a path */
  search(path: string): Format | undefined {
    for (const format of this.entries.values()) {
      if (path.endsWith(format.ext)) {
        return format;
      }
    }
  }

  /** Delete a cached template */
  deleteCache(file: string): void {
    for (const format of this.entries.values()) {
      format.engines?.forEach((engine) => engine.deleteCache(file));
    }
  }
}
