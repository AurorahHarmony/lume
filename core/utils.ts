import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import { brightGreen, cyan, dim, gray, green, red } from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "feed",
  "filter_pages",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "mdx",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nav",
  "decap_cms",
  "on_demand",
  "pagefind",
  "picture",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "sitemap",
  "slugify_urls",
  "source_maps",
  "svgo",
  "tailwindcss",
  "terser",
  "toml",
  "unocss",
  "vento",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.33.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  const command = "deno task lume upgrade" + (stable ? "" : " --dev");

  log(
    `Update available ${dim(current)} â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(
  branch = "master",
): Promise<string> {
  const response = await fetch(
    `https://api.github.com/repos/lumeland/lume/commits/${branch}`,
  );
  const commits = await response.json();
  return commits.sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement &&
    // @ts-ignore: Check if the argument passed is a Page.data object
    obj !== obj.page?.data;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string, rootToRemove?: string) {
  if (rootToRemove) {
    path = path.replace(rootToRemove, "");
  }

  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      if (rootToRemove && path.startsWith(rootToRemove)) {
        return posix.join("/", path.replace(rootToRemove, ""));
      }

      return path;
    }
  }

  const absolute = posix.join("/", path);
  return rootToRemove && absolute.startsWith(rootToRemove)
    ? posix.join("/", absolute.replace(rootToRemove, ""))
    : absolute;
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch {
      throw new Error(`Config file not found (${path})`);
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports as Record<string, string>,
          scopes: config.scopes as Record<string, Record<string, string>>,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

export function updateLumeVersion(url: URL, denoConfig: DenoConfigResult) {
  denoConfig.importMap ??= { imports: {} };

  const { config, importMap } = denoConfig;

  // Configure the import map
  if (Deno.version.deno < "1.30.0") {
    config.importMap ||= "./import_map.json";
  }

  const oldUrl = importMap.imports["lume/"];
  const newUrl = new URL("./", url).href;
  importMap.imports["lume/"] = newUrl;

  for (const [specifier, url] of Object.entries(importMap.imports)) {
    if (url.startsWith(oldUrl)) {
      importMap.imports[specifier] = url.replace(oldUrl, newUrl);
    }
  }

  // Configure lume tasks
  const tasks = config.tasks || {};
  if (!tasks.lume || !tasks.lume.includes(`echo "import 'lume/cli.ts'"`)) {
    tasks.lume = `echo "import 'lume/cli.ts'" | deno run --unstable -A -`;
    tasks.build = "deno task lume";
    tasks.serve = "deno task lume -s";
  }
  config.tasks = tasks;
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.imports = importMap.imports;
    config.scopes = importMap.scopes;
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    const save = confirm(
      "Saving the deno.jsonc file will overwrite the comments. Continue?",
    );

    if (!save) {
      console.log(
        "You have to update your deno.jsonc file manually with the following content:",
      );
      console.log(dim(JSON.stringify(config, null, 2)));
      console.log("Use deno.json to update it automatically without asking.");
      return;
    }
  }
  await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
  console.log("Deno configuration file saved:", gray(file));
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}
export function isAbsolutePath(path: string): boolean {
  return SEP !== "/" ? /^\w:[\/\\]/.test(path) : path.startsWith("/");
}

export function replaceExtension(
  path: string | false,
  ext: string,
): string | false {
  if (!path) {
    return false;
  }
  return path.replace(/\.\w+$/, ext);
}

export function getPathAndExtension(path: string): [string, string] {
  const extension = getExtension(path);
  const pathWithoutExtension = path.slice(0, -extension.length);
  return [pathWithoutExtension, extension];
}

export function getExtension(path: string): string {
  const match = path.match(/\.\w+$/);
  return match ? match[0] : "";
}

export async function read(
  path: string,
  isBinary: boolean,
): Promise<Uint8Array | string>;
export async function read(
  path: string,
  isBinary: true,
  init?: RequestInit,
): Promise<Uint8Array>;
export async function read(
  path: string,
  isBinary: false,
  init?: RequestInit,
): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
  init?: RequestInit,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    if (path.startsWith("data:")) {
      const response = await fetch(path);

      return isBinary
        ? new Uint8Array(await response.arrayBuffer())
        : response.text();
    }

    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const url = new URL(path);

  if (url.protocol === "file:") {
    return isBinary ? Deno.readFile(url) : Deno.readTextFile(url);
  }

  const cache = await caches.open("lume_remote_files");

  // Prevent https://github.com/denoland/deno/issues/19696
  try {
    const cached = await cache.match(url);

    if (cached) {
      return isBinary
        ? new Uint8Array(await cached.arrayBuffer())
        : cached.text();
    }
  } catch {
    // ignore
  }

  const response = await fetch(url, init);
  await cache.put(url, response.clone());

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}

/**
 * Check if the content variable is a generator.
 */
export function isGenerator(
  content: unknown,
): content is GeneratorFunction | AsyncGeneratorFunction {
  if (typeof content !== "function") {
    return false;
  }

  const name = content.constructor.name;
  return (name === "GeneratorFunction" || name === "AsyncGeneratorFunction");
}

/**
 * Resolve the path of an included file
 * Used in the template engines and processors
 */
export function resolveInclude(
  path: string,
  includesDir: string,
  fromDir?: string,
  rootToRemove?: string,
): string {
  if (isUrl(path)) {
    return path;
  }

  if (path.startsWith(".")) {
    if (!fromDir) {
      throw new Error(`Cannot load "${path}" without a base directory`);
    }

    return normalizePath(posix.join(fromDir, path), rootToRemove);
  }

  const normalized = normalizePath(path, rootToRemove);

  return normalized.startsWith(normalizePath(posix.join(includesDir, "/")))
    ? normalized
    : normalizePath(posix.join(includesDir, normalized));
}

export function env<T>(name: string): T | undefined {
  const value = Deno.env.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

export function subExtensions(extensions: string[], subextension?: string) {
  if (!subextension) {
    return extensions;
  }
  if (!subextension.startsWith(".")) {
    throw new Error(
      `Invalid subextension: "${subextension}". It must start with "."`,
    );
  }
  return extensions.map((ext) => subextension + ext);
}
