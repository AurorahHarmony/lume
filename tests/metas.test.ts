import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import metas from "../plugins/metas.ts";
import { Page } from "../core.ts";

Deno.test("metas plugin", async (t) => {
  const site = getSite({
    src: "metas",
  });

  site.use(metas());
  site.preprocess([".md"], (page: Page) => {
    page.data.excerpt ??= (page.data.content as string).split("<!--more-->")[0];
  });

  await build(site);
  await assertSiteSnapshot(t, site);
});
