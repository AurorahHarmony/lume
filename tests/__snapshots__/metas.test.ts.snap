export const snapshot = {};

snapshot[`metas plugin 1`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.jsonc",
      pageLoader: [AsyncFunction: json],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".jsonc",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: undefined,
      ext: ".toml",
      pageLoader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
  src: [
    "/",
    "/_data.yml",
    "/page-1.njk",
    "/page-2.njk",
    "/page-3.md",
  ],
}
`;

snapshot[`metas plugin 2`] = `[]`;

snapshot[`metas plugin 3`] = `
[
  {
    content: '<!DOCTYPE html>
<html><head>
  <meta property="og:type" content="website">
<meta property="og:site_name" content="My site">
<meta property="og:locale" content="gl">
<meta property="og:title" content="Custom title">
<meta property="og:description" content="A very long custom description with a lot of text that will be truncated to a maximum of 155 characters and then append &quot;…&quot; to the end. Lorem ipsum dolor…">
<meta property="og:url" content="http://localhost/page-1/">
<meta property="og:image" content="http://localhost/my-image.png">
<meta name="twitter:title" content="Custom title">
<meta name="twitter:description" content="A very long custom description with a lot of text that will be truncated to a maximum of 155 characters and then append &quot;…&quot; to the end. Lorem ipsum dolor sit amet, consectetur adipiscing elit.">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="http://localhost/my-image.png">
<meta name="twitter:site" content="@myUser">
<meta itemprop="name" content="Custom title">
<meta itemprop="description" content="A very long custom description with a lot of text that will be truncated to a maximum of 155 characters and then append &quot;…&quot; to the end. Lorem ipsum dolor…">
<meta itemprop="image" content="http://localhost/my-image.png">
<meta name="description" content="A very long custom description with a lot of text that will be truncated to a maximum of 155 characters and then append &quot;…&quot; to the end. Lorem ipsum dolor…">
<meta name="keywords" content="one, two">
<meta name="robots" content="index, follow">
<meta name="theme-color" content="black">
<meta name="generator" content="Lume testing">
</head>
  <body>
    Hello world
  

</body></html>',
    data: {
      children: "<html>
  <head>
  </head>
  <body>
    Hello world
  </body>
</html>
",
      content: "<html>
  <head>
  </head>
  <body>
    Hello world
  </body>
</html>
",
      date: [],
      mergedKeys: [
        "tags",
        "metas",
      ],
      metas: [
        "site",
        "title",
        "description",
        "image",
        "icon",
        "robots",
        "keywords",
        "twitter",
        "lang",
        "color",
        "generator",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page-1/",
    },
    src: {
      asset: undefined,
      ext: ".njk",
      path: "/page-1",
      remote: undefined,
      slug: "page-1",
    },
  },
  {
    content: '<!DOCTYPE html>
<html><head>
  <meta property="og:type" content="website">
<meta property="og:site_name" content="My site">
<meta property="og:locale" content="gl">
<meta property="og:title" content="Relative paths">
<meta property="og:description" content="Tests the use of relative path (to page.data.url) when filling out the og:image or og:icon URL">
<meta property="og:url" content="http://localhost/page-2/">
<meta property="og:image" content="http://localhost/page-2/my-image.png">
<meta name="twitter:title" content="Relative paths">
<meta name="twitter:description" content="Tests the use of relative path (to page.data.url) when filling out the og:image or og:icon URL">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="http://localhost/page-2/my-image.png">
<meta name="twitter:site" content="@myUser">
<meta itemprop="name" content="Relative paths">
<meta itemprop="description" content="Tests the use of relative path (to page.data.url) when filling out the og:image or og:icon URL">
<meta itemprop="image" content="http://localhost/page-2/my-image.png">
<meta name="description" content="Tests the use of relative path (to page.data.url) when filling out the og:image or og:icon URL">
<meta name="keywords" content="one, two">
<meta name="robots" content="noindex, nofollow, noarchive">
<meta name="theme-color" content="black">
<meta name="generator" content="Lume testing">
</head>
  <body>
    Hello world
  

</body></html>',
    data: {
      children: "<html>
  <head>
  </head>
  <body>
    Hello world
  </body>
</html>
",
      content: "<html>
  <head>
  </head>
  <body>
    Hello world
  </body>
</html>
",
      date: [],
      mergedKeys: [
        "tags",
        "metas",
      ],
      metas: [
        "site",
        "title",
        "description",
        "image",
        "icon",
        "robots",
        "keywords",
        "twitter",
        "lang",
        "color",
        "generator",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page-2/",
    },
    src: {
      asset: undefined,
      ext: ".njk",
      path: "/page-2",
      remote: undefined,
      slug: "page-2",
    },
  },
  {
    content: '<!DOCTYPE html>
<html><head><meta property="og:type" content="website">
<meta property="og:site_name" content="My site">
<meta property="og:locale" content="gl">
<meta property="og:title" content="Title from page data">
<meta property="og:description" content="This is page excerpt will be used as meta description.">
<meta property="og:url" content="http://localhost/page-3/">
<meta property="og:image" content="http://localhost/page-3/use-cover-as-meta-image.png">
<meta name="twitter:title" content="Title from page data">
<meta name="twitter:description" content="This is page excerpt will be used as meta description.">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="http://localhost/page-3/use-cover-as-meta-image.png">
<meta name="twitter:site" content="@myUser">
<meta itemprop="name" content="Title from page data">
<meta itemprop="description" content="This is page excerpt will be used as meta description.">
<meta itemprop="image" content="http://localhost/page-3/use-cover-as-meta-image.png">
<meta name="description" content="This is page excerpt will be used as meta description.">
<meta name="keywords" content="one, two">
<meta name="robots" content="this robots will be overrided">
<meta name="theme-color" content="black">
<meta name="generator" content="Lume testing">
</head><body><p>This is page excerpt will be used as meta description.</p>
<!--more-->
</body></html>',
    data: {
      children: "<p>This is page excerpt will be used as meta description.</p>
<!--more-->
",
      content: "This is page excerpt will be used as meta description.

<!--more-->
",
      cover: "./use-cover-as-meta-image.png",
      date: [],
      excerpt: "This is page excerpt will be used as meta description.

",
      header: [
        "title",
      ],
      mergedKeys: [
        "tags",
        "metas",
      ],
      metas: [
        "site",
        "title",
        "description",
        "image",
        "icon",
        "robots",
        "keywords",
        "twitter",
        "lang",
        "color",
        "generator",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      robots: "this robots will be overrided",
      search: [],
      tags: "Array(0)",
      url: "/page-3/",
    },
    src: {
      asset: undefined,
      ext: ".md",
      path: "/page-3",
      remote: undefined,
      slug: "page-3",
    },
  },
]
`;
